File: 501-find-mode-in-binary-search-tree/501.py
class Solution:
    def dfs(self, root):
        if root is None:
            return

        if root.val not in self.table:
            self.table[root.val] = 1
        else:
            self.table[root.val] += 1

        self.dfs(root.left)
        self.dfs(root.right)

    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.table = {}
        self.dfs(root)
        sort = sorted(self.table.items(), key=lambda x:x[1], reverse=True)

        ans = []
        Max = -math.inf
        for k,v in sort:
            if v >= Max:
                ans.append(k)
                Max = v

        return ans

File: 501-find-mode-in-binary-search-tree/501.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    map<int, int> m;
    int count;
    void dfs(TreeNode* node) {
        if (!node) {
            return;
        }
        dfs(node->left);
        ++m[node->val];
        count = max(count, m[node->val]);
        dfs(node->right);
    }
public:
    vector<int> findMode(TreeNode* root) {
        count = 0;
        dfs(root);
        vector<int> ans;
        for (const pair<int, int>& p : m) {
            if (p.second == count) {
                ans.push_back(p.first);
            }
        }
        return ans;
    }
};
